kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/RecordEncryptionST.java

```diff
@@ -115,24 +128,114 @@ void produceAndConsumeMessage(String namespace, TestKmsFacade<?, ?, ?> testKmsFa
+        KroxyliciousSteps.produceMessages(namespace, topicName, bootstrap, MESSAGE, numberOfMessages);
+
+        LOGGER.info("Then the messages are consumed");
+        List<ConsumerRecord> resultEncryptedRotatedKek = KroxyliciousSteps.consumeMessageFromKafkaCluster(namespace, topicName, clusterName,
+                Constants.KAFKA_DEFAULT_NAMESPACE, numberOfMessages, Duration.ofMinutes(2));
+        LOGGER.info("Received: {}", resultEncryptedRotatedKek);
+
+        assertThat(resultEncryptedRotatedKek.stream())
+                .withFailMessage("v2 is not contained in the ciphertext blob!")
+                .allMatch(r -> testKekManager.assertKekVersionWithinParcel(r.getValue()));
+    }
+
+    @TestTemplate
+    void produceAndConsumeMessageWithRotatedKEK(String namespace, TestKmsFacade<?, ?, ?> testKmsFacade) {
+        testKekManager = testKmsFacade.getTestKekManager();
+        testKekManager.generateKek("KEK_" + topicName);
+        int numberOfMessages = 1;
+        ExperimentalKmsConfig experimentalKmsConfig = new ExperimentalKmsConfig(1, 1, 1, 1);
+
+        // start Kroxylicious
+        LOGGER.info("Given Kroxylicious in {} namespace with {} replicas", namespace, 1);
+        Kroxylicious kroxylicious = new Kroxylicious(namespace);
+        kroxylicious.deployPortPerBrokerPlainWithRecordEncryptionFilter(clusterName, 1, testKmsFacade, experimentalKmsConfig);
+        bootstrap = kroxylicious.getBootstrap();
+
+        LOGGER.info("And a kafka Topic named {}", topicName);
+        KafkaSteps.createTopic(namespace, topicName, bootstrap, 1, 2);
+
+        LOGGER.info("When {} messages '{}' are sent to the topic '{}'", numberOfMessages, MESSAGE, topicName);
+        KroxyliciousSteps.produceMessages(namespace, topicName, bootstrap, MESSAGE, numberOfMessages);
+
+        LOGGER.info("Then the messages are consumed");
+        List<ConsumerRecord> result = KroxyliciousSteps.consumeMessages(namespace, topicName, bootstrap, numberOfMessages, Duration.ofMinutes(2));
+        LOGGER.info("Received: {}", result);
+
+        assertThat(result).withFailMessage("expected messages have not been received!")
+                .extracting(ConsumerRecord::getValue)
+                .hasSize(numberOfMessages)
+                .allSatisfy(v -> assertThat(v).contains(MESSAGE));
+
+        LOGGER.info("When KEK is rotated");
+        testKekManager.rotateKek("KEK_" + topicName);
+
+        LOGGER.info("And {} messages '{}' are sent to the topic '{}'", numberOfMessages, MESSAGE, topicName);
+        KroxyliciousSteps.produceMessages(namespace, topicName, bootstrap, MESSAGE, numberOfMessages);
+
+        LOGGER.info("Then the messages are consumed");
+        List<ConsumerRecord> resultRotatedKek = KroxyliciousSteps.consumeMessages(namespace, topicName, bootstrap, numberOfMessages, Duration.ofMinutes(2));
+        LOGGER.info("Received: {}", resultRotatedKek);
+
+        assertThat(resultRotatedKek).withFailMessage("expected messages have not been received!")
```

Comment:
I think the semantics of `KafkaClient#consumeMessages` is to consume x messages from the topic, starting from the beginning on the topic. No?

Assuming that is true, the test is incorrect. `resultRotatedKek` will have both the original messages and the newly sent ones.

To fix this you have a choice:
* Add a new `consumeMessages` that uses a fixed consumer group.    You'd need to check that all the APIs support the specification of a consumer group.
* Change this assertion to ignore the messages that it received earlier so that it only considers the new messages.
*